File: 24-Graph-Part-II/java/06-MstUsingKruskalsAlgo.java
import java.util.*;

class Solution {
    public int spanningTree(int V, List<List<List<Integer>>> adj) {
        List<int[]> edges = new ArrayList<>();
        for (int u = 0; u < V; u++) {
            for (List<Integer> neighbor : adj.get(u)) {
                int v = neighbor.get(0);
                int weight = neighbor.get(1);
                if (u < v) {
                    edges.add(new int[]{weight, u, v});
                }
            }
        }

        edges.sort((a, b) -> a[0] - b[0]);

        int[] parent = new int[V];
        int[] rank = new int[V];
        for (int i = 0; i < V; i++) {
            parent[i] = i;
        }

        int sum = 0;
        for (int[] edge : edges) {
            int weight = edge[0];
            int u = edge[1];
            int v = edge[2];
            int pu = find(u, parent);
            int pv = find(v, parent);
            if (pu != pv) {
                sum += weight;
                union(pu, pv, parent, rank);
            }
        }

        return sum;
    }

    private int find(int node, int[] parent) {
        if (parent[node] != node) {
            parent[node] = find(parent[node], parent);
        }
        return parent[node];
    }

    private void union(int u, int v, int[] parent, int[] rank) {
        if (rank[u] < rank[v]) {
            parent[u] = v;
        } else if (rank[u] > rank[v]) {
            parent[v] = u;
        } else {
            parent[v] = u;
            rank[u]++;
        }
    }
}

File: 24-Graph-Part-II/python/06_mst_using_kruskals_algo.py
from typing import List

class Solution:
    def spanningTree(self, V: int, adj: List[List[List[int]]]) -> int:
        edges = []
        for u in range(V):
            for neighbor in adj[u]:
                v, weight = neighbor[0], neighbor[1]
                if u < v:
                    edges.append((weight, u, v))

        edges.sort()

        parent = list(range(V))
        rank = [0] * V

        def find(node):
            if parent[node] != node:
                parent[node] = find(parent[node])
            return parent[node]

        def union(u, v):
            if rank[u] < rank[v]:
                parent[u] = v
            elif rank[u] > rank[v]:
                parent[v] = u
            else:
                parent[v] = u
                rank[u] += 1

        total_sum = 0
        for weight, u, v in edges:
            pu, pv = find(u), find(v)
            if pu != pv:
                total_sum += weight
                union(pu, pv)

        return total_sum
